src/template_partials/templatetags/partials.py

```diff
@@ -142,34 +142,28 @@ def _define_partial(parser, token, end_tag):
     return DefinePartialNode(partial_name, inline, nodelist)
 
 
-class SubDictionaryWrapper:
+class PartialContentsWrapper:
     """
-    Wrap a parent dictionary, allowing deferred access to a sub-dictionary by key.
-
-    The parser.extra_data storage may not yet be populated when a partial node
-    is defined, so we defer access until rendering.
+    Wrapper for accessing partial contents either from parser.extra_data or parser.origin.
+    Handles deferred access to partial contents, which may not be populated when a partial node is defined.
     """
 
-    def __init__(self, parent_dict, lookup_key):
+    def __init__(self, parent_dict=None, lookup_key=None, origin=None):
         self.parent_dict = parent_dict
         self.lookup_key = lookup_key
-
-    def __getitem__(self, key):
-        return self.parent_dict[self.lookup_key][key]
-
-
-class OriginPartialContentsWrapper:
-    """
-    Wrap parser.origin to allow deferred access to partial_contents.
-    """
-
-    def __init__(self, origin):
         self.origin = origin
 
     def __getitem__(self, key):
-        if not hasattr(self.origin, "partial_contents"):
-            self.origin.partial_contents = {}
-        return self.origin.partial_contents[key]
+        try:
+            if self.origin is None:
+                return self.parent_dict[self.lookup_key][key]
+            else:
+                if not hasattr(self.origin, "partial_contents"):
+                    self.origin.partial_contents = {}
+                return self.origin.partial_contents[key]
+
+        except KeyError:
+            raise template.TemplateSyntaxError(f"Undefined partial: '{key}'")
```

Comment:
Please separate this into two changes: the refactoring and the improvement of the error message. 

For the refactoring, assume the 5.1+ usage, and catch the type error to fallback. For the improvement, catch only the second dict access. (If the first one fails we have other problems.

Untested, by something like: 

```python
    try: 
        partials_content = self.parent_dict[self.lookup_key]
    except TypeError:
        # Pre Django 5.1 storage.
        partials_content = getattr(self.parent_dict, self.lookup_key, {})
        
    try: 
        return partials_content[key]
    except KeyError:
        raise template.TemplateSyntaxError(f"Undefined partial: '{key}'")        
```

(I'm not sure about the _syntax_ error there, but I'll have a think about that. 🤔)

Reply:
Is that because no partials at all are defined? (What happens if you define a different partial?) 

We should raise a different error there. (Likely someone hits it moving a partial from one template to another. We should try to be as helpful as possible.

Reply:
In the test, the origin is <unknown source>, so that is why the tests are failing. I think .

Reply:
I have implemented the EAFP method to handle versions.